SharkCage/SharedFunctionality/ValidateBinary.cpp

```diff
@@ -0,0 +1,226 @@
+#include "stdafx.h"
+#include "ValidateBinary.h"
+
+#include <tchar.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <windows.h>
+#include <Softpub.h>
+#include <wincrypt.h>
+#include <wintrust.h>
+#include <fstream>
+
+#pragma comment(lib, "wintrust")
+#pragma comment(lib, "bcrypt.lib")
+
+#define NT_SUCCESS(Status)          (((NTSTATUS)(Status)) >= 0)
+#define STATUS_UNSUCCESSFUL         ((NTSTATUS)0xC0000001L)
+
+bool ValidateBinary::ValidateCertificate(const std::wstring &app_path)
+{
+	WINTRUST_FILE_INFO FileData = { 0 };
+	FileData.cbStruct = sizeof(WINTRUST_FILE_INFO);
+	FileData.pcwszFilePath = app_path.c_str();
+	FileData.hFile = nullptr;
+	FileData.pgKnownSubject = nullptr;
+
+	WINTRUST_DATA WinTrustData = { 0 };
+	WinTrustData.cbStruct = sizeof(WinTrustData);
+	WinTrustData.pPolicyCallbackData = nullptr;
+	WinTrustData.pSIPClientData = nullptr;
+	WinTrustData.dwUIChoice = WTD_UI_NONE;
+	WinTrustData.fdwRevocationChecks = WTD_REVOKE_NONE;
+	WinTrustData.dwUnionChoice = WTD_CHOICE_FILE;
+	WinTrustData.dwStateAction = WTD_STATEACTION_VERIFY;
+	WinTrustData.hWVTStateData = nullptr;
+	WinTrustData.pwszURLReference = nullptr;
+	WinTrustData.dwUIContext = 0;
+	WinTrustData.pFile = &FileData;
+
+	GUID WVTPolicyGUID = WINTRUST_ACTION_GENERIC_VERIFY_V2;
+
+	LONG status = ::WinVerifyTrust(
+		nullptr,
+		&WVTPolicyGUID,
+		&WinTrustData);
+
+	WinTrustData.dwStateAction = WTD_STATEACTION_CLOSE;
+
+	return status == ERROR_SUCCESS;
+}
+
+bool ValidateBinary::ValidateHash(const std::wstring &app_path, const std::wstring &app_hash)
+{
+	//open file
+	std::ifstream infile(app_path, std::ios::binary);
+	std::vector<char> buffer;
+
+	//get length of file
+	infile.seekg(0, infile.end);
+	size_t length = static_cast<size_t>(infile.tellg());
+	infile.seekg(0, infile.beg);
+
+	//read file
+	if (length < 1) {
+		return false;
+	}
+
+	buffer.resize(length);
+	infile.read(&buffer[0], length);
+
+	BCRYPT_ALG_HANDLE algorithm = nullptr;
+	BCRYPT_HASH_HANDLE hash_handle = nullptr;
+	NTSTATUS status = STATUS_UNSUCCESSFUL;
+	DWORD data = 0, cb_hash = 0, cb_hash_object = 0;
+	PBYTE hash_object = nullptr;
```

Comment:
this should also be `PUCHAR` according to the crypt functions documentation, after that, the `PBYTE` casts should no longer be necessary (or at least possible with static_cast)